ui/jobs/jobs.js

```diff
@@ -2815,13 +2851,107 @@ class Bars {
   }
 
   setupNin() {
+    const ninki = this.addResourceBox({
+      classList: ['nin-color-ninki'],
+    });
+    const hutonBox = this.addProcBox({
+      id: 'nin-procs-huton',
+      fgColor: 'nin-color-huton',
+      threshold: 20,
+    });
+    const trickAttack = this.addProcBox({
+      id: 'nin-procs-trickattack',
+      fgColor: 'nin-color-trickattack',
+    });
+    const bunshin = this.addProcBox({
+      id: 'nin-procs-bunshin',
+      fgColor: 'nin-color-bunshin',
+    });
+    this.abilityFuncMap[kAbility.Bunshin] = () => {
+      bunshin.duration = 0;
+      bunshin.duration = 90;
+    };
+    const ninjutsu = this.addProcBox({
+      id: 'nin-procs-ninjutsu',
+      fgColor: 'nin-color-ninjutsu',
+    });
+    // Ninjutsu's cooldown begins to countdown at the first mudra.
+    this.gainEffectFuncMap[EffectId.Mudra] = () => {
+      if (this.mudraTriggerCd === false)
+        return;
+      let old = parseFloat(ninjutsu.duration) - parseFloat(ninjutsu.elapsed);
+      if (old > 0) {
+        ninjutsu.duration = 0;
+        ninjutsu.duration = old + 20;
+      } else {
+        ninjutsu.duration = 0;
+        ninjutsu.duration = 20 - 0.5;
+      }
+      this.mudraTriggerCd = false;
+    };
+    // On each mudra, Mudra effect will be gain once,
+    // use this.mudraTriggerCd to tell that whether this mudra trigger cooldown.
+    this.loseEffectFuncMap[EffectId.Mudra] = () => {
+      this.mudraTriggerCd = true;
+    };
+    this.gainEffectFuncMap[EffectId.Kassatsu] = () => {
+      this.mudraTriggerCd = false;
+    };
+    this.loseEffectFuncMap[EffectId.Kassatsu] = () => {
+      this.mudraTriggerCd = true;
+    };
+    this.abilityFuncMap[kAbility.Hide] = () => {
+      ninjutsu.duration = 0;
+    };
+    this.statChangeFuncMap['NIN'] = () => {
+      trickAttack.valuescale = this.gcdSkill();
+      this.abilityFuncMap[kAbility.TrickAttack] = () => {
+        trickAttack.duration = 0;
+        trickAttack.duration = 15;
+        trickAttack.threshold = 1000;
+        trickAttack.fg = computeBackgroundColorFrom(trickAttack, 'nin-color-trickattack.active');
+        setTimeout(() => {
+          trickAttack.duration = 45;
+          trickAttack.threshold = this.gcdSkill() * 4;
+          trickAttack.fg = computeBackgroundColorFrom(trickAttack, 'nin-color-trickattack');
+        }, 15000);
+      };
+      bunshin.valuescale = this.gcdSkill();
+      bunshin.threshold = this.gcdSkill() * 8;
+      ninjutsu.valuescale = this.gcdSkill();
+      ninjutsu.threshold = this.gcdSkill() * 2;
+    };
+
     this.jobFuncs.push((jobDetail) => {
       if (jobDetail.hutonMilliseconds > 0) {
         if (this.huton != 1)
           this.huton = 1;
       } else if (this.huton == 1) {
         this.huton = 0;
```

Comment:
what is `this.huton` use for ?

Reply:
it has been there when I write. I think it is for GCD calculation?

Reply:
Sorry, I saw the code in the wrong place. (ーー;)

Reply:
This is used for exactly one thing, as a boolean. It would probably be okay to just rewrite the body of the function as `this.huton = !this.huton` instead of checking numeric values.  But it might be something @quisquous would prefer to do in a separate pull request.